backend/push.js

```diff
@@ -1,53 +1,292 @@
-    socket.server.pushSubscriptions[subscription.endpoint] = subscription
+    if (!server.pushSubscriptions[subscriptionId]) {
+      server.pushSubscriptions[subscriptionId] = subscriptionInfoWrapper(subscriptionId, subscription)
+      addSubscriptionToIndex(subscriptionId).then(() => {
+        return sbp('chelonia/db/set', `_private_webpush_${subscriptionId}`, JSON.stringify({ subscription: subscription, channelIDs: [] }))
+      }).catch((e) => console.error(e, 'Error saving subscription', subscriptionId))
+      postEvent(server.pushSubscriptions[subscriptionId], JSON.stringify({ type: 'initial' })).catch(e => console.warn(e, 'Error sending initial push notification'))
+    } else {
+      if (server.pushSubscriptions[subscriptionId].sockets.size === 0) {
+        server.pushSubscriptions[subscriptionId].subscriptions.forEach((channelID) => {
+          if (!server.subscribersByChannelID[channelID]) return
+          server.subscribersByChannelID[channelID].delete(server.pushSubscriptions[subscriptionId])
+        })
+      }
+    }
+    if (socket.pushSubscriptionId) {
+      if (socket.pushSubscriptionId === subscriptionId) return
+      const oldSubscriptionId = socket.pushSubscriptionId
+      server.pushSubscriptions[oldSubscriptionId].sockets.delete(socket)
+      if (server.pushSubscriptions[oldSubscriptionId].sockets.size === 0) {
+        server.pushSubscriptions[oldSubscriptionId].subscriptions.forEach((channelID) => {
+          if (!server.subscribersByChannelID[channelID]) {
+            server.subscribersByChannelID[channelID] = new Set()
+          }
+          server.subscribersByChannelID[channelID].add(server.pushSubscriptions[oldSubscriptionId])
+        })
+      }
+    }
+    socket.pushSubscriptionId = subscriptionId
+    server.pushSubscriptions[subscriptionId].subscriptions.forEach((channelID) => {
+      server.subscribersByChannelID?.[channelID].delete(server.pushSubscriptions[subscriptionId])
+    })
+    server.pushSubscriptions[subscriptionId].sockets.add(socket)
+    socket.subscriptions?.forEach(channelID => {
+      server.pushSubscriptions[subscriptionId].subscriptions.add(channelID)
+    })
+    saveSubscription(server, subscriptionId)
   },
-  [PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION] (payload) {
+  [PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION] () {
     const socket = this
-    const subscriptionId = JSON.parse(payload)
+    const { server, pushSubscriptionId: subscriptionId } = socket
 
-    delete socket.server.pushSubscriptions[subscriptionId]
-  },
-  [PUSH_SERVER_ACTION_TYPE.SEND_PUSH_NOTIFICATION]: async function (payload) {
-    const pushSubscriptions = this.server.pushSubscriptions
-    const data = JSON.parse(payload)
-    const sendPush = (sub) => pushController.sendNotification(
-      sub, JSON.stringify({ title: data.title, body: data.body })
-    )
+    if (subscriptionId) {
+      removeSubscription(server, subscriptionId)
+    }
+  }
+}
```

Comment:
It is somewhat frustrating for me to read this code for some of these referenced functions (like `saveSubscription`) are above this chunk of handlers, and others (`removeSubscription`) are below.

Could the functions be kept with the functions, and the handlers with the handlers? I.e. could you move the handlers down to the bottom of the file (or at the top)?